warehouse/rate_limiting/__init__.py

```diff
@@ -162,7 +163,82 @@ def __eq__(self, other):
         )
 
 
+def ratelimit_view_deriver(view, info):
+    """
+    A general-purpose rate limit view deriver based on client's `remote_addr`.
+    https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/hooks.html#custom-view-derivers
+
+    Add `ratelimiter="<limit-string>"` to a `@view_config()`. Example:
+
+    ```
+    @view_config(
+        ...,
+        ratelimit="10/second",
+        ratelimit_by="user.id",  # Optional, defaults to "remote_addr"
+    )
+    ```
+    https://limits.readthedocs.io/en/stable/quickstart.html#rate-limit-string-notation
+    """
+    limit_str = info.options.get("ratelimit")
+    limit_by = info.options.get("ratelimit_by", "remote_addr")
+    if limit_str:
+
+        def wrapper_view(context, request):
+            limit_attr = getattr(request, limit_by)
+
+            ratelimiter = request.find_service(
+                IRateLimiter, name="rate_limiting.client"
+            )
+            # TODO: Is this Kosher?
+            # Override the default limits with the ones specified in the view
+            ratelimiter._limits = parse_many(limit_str)
+
+            metrics = request.find_service(IMetricsService, context=None)
+
+            request_route = request.matched_route.name
+
+            ratelimiter.hit(limit_attr)
+            if not ratelimiter.test(limit_attr):
+                metrics.increment(
+                    "warehouse.ratelimiter.exceeded",
+                    tags=[
+                        f"request_route:{request_route}",
+                    ],
+                )
+                message = (
+                    "The action could not be performed because there were too "
+                    "many requests by the client."
+                )
+                _resets_in = ratelimiter.resets_in(limit_attr)
+                if _resets_in is not None:
+                    _resets_in = max(1, int(_resets_in.total_seconds()))
+                    message += f" Limit may reset in {_resets_in} seconds."
+                raise HTTPTooManyRequests(message)
+
+            metrics.increment(
+                "warehouse.ratelimiter.hit",
+                tags=[
+                    f"request_route:{request_route}",
+                ],
+            )
+            return view(context, request)
+
+        return wrapper_view
+    return view
+
+
+ratelimit_view_deriver.options = {"ratelimit", "ratelimit_by"}  # type: ignore[attr-defined] # noqa: E501
+
+
 def includeme(config):
     config.registry["ratelimiter.storage"] = storage_from_string(
         config.registry.settings["ratelimit.url"]
     )
+
+    config.register_service_factory(
+        RateLimit(limit="100 per second"),
```

Comment:
I also couldn't figure out a way to not pass a `limit` arg here, so I picked something. Since any users of the view deriver have to override the string anyhow, is that a problem?